### ISSUE TITLE
Realm not working correctly when creating Class in main thread and use immediately in background thread

### ISSUE THREAD

**Issue by mihirjoshi21** [realm/realm-java#3037]
#### Goal

> I have a Rest API which works after getting the AccessToken from the server. I API failed or if the access token expires, I fetch the token and save it in realm database using token API and restart the same request. Access Token is saved in the main thread and immediately accessed in the background thread in OKHttp Interceptor.

#### Expected Results

>  AccessToken should work immediately after saving.

#### Actual Results 

>  AccessToken is not working for the first time immediately after saving the token but after closing the application and reopening it is working.

#### Steps & Code to Reproduce

>  

After fetching token -
```
Realm realm = Realm.getDefaultInstance();
        realm.executeTransactionAsync(bgRealm -> {
            bgRealm.copyToRealm(accessToken);
        }, () -> {
            mHomePresenter.loadRepositories(accessToken, "280", "1038");
            realm.close();
        }, error -> {
            error.getMessage();
        });
```

and then using it for loading Repositories.

```
Interceptor interceptor = chain -> {
                AccessToken token = Realm.getDefaultInstance().where(AccessToken.class).findFirst();
                Request newRequest = chain.request().newBuilder().addHeader("Authorization",
                        "Bearer " + (token != null ? token.getTokenInfo().getAccessToken() : "")).build();
                return chain.proceed(newRequest);
            };
```

#### Version of Realm and tooling
Realm version(s): ? 1.0.0

Android Studio version: ? 2.1.1

Which Android version and device: ? Android M

**Comment by mihirjoshi21** [comment]
`loadRepositories() ` is the Retrofit Observable pattern.

```
crmnextService.apiHomePage(ownerId, roleId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(application.defaultSubscribeScheduler())
                .doOnError(throwable -> {
                    if (isHttp401(throwable)) {
                        fetchAuthToken(accessToken, mHomeMvpView);
                    }
                })
                .subscribe(new Subscriber<List<HomeItem>>() {
                    @Override
                    public void onCompleted() {
                            // Save Realm Data
                            mHomeMvpView.showHome(mHomeItems);
                        } else {
                            // Empty repo
                        }
                    }

                    @Override
                    public void onError(Throwable error) {
                        if (isHttp401(error)) {
                        } else{
                              \\ Fetch Saved Realm Data
                        }
                    }

                    @Override
                    public void onNext(List<HomeItem> homeItems) {
                        mHomeItems = homeItems;
                    }
                });
```

**Comment by cmelchior** [comment]
It sounds like it might have the same root cause as https://github.com/realm/realm-java/issues/1893. Looking into it now.

**Comment by beeender** [comment]
@mihirjoshi21 BTW, `AccessToken token = Realm.getDefaultInstance().where(AccessToken.class).findFirst();` might cause leaks since the Realm instance returned by `getDefaultInstance()` is not properly closed.

**Comment by mihirjoshi21** [comment]
@beeender I will close it eventually. The strange part is after saving if I force close the application and then restart it everything works just fine. Another thing which I forgot to mention is if I fetch token outside Interceptor it is getting the data but then I am getting Realm used in the incorrect thread.

```
 AccessToken token = Realm.getDefaultInstance().where(AccessToken.class).findFirst();
Interceptor interceptor = chain -> {
                Request newRequest = chain.request().newBuilder().addHeader("Authorization",
                        "Bearer " + (token != null ? token.getTokenInfo().getAccessToken() : "")).build();
                return chain.proceed(newRequest);
            };
```

**Comment by Zhuinden** [comment]
You can't close it, you don't have a reference to it.

**Comment by kneth** [comment]
@mihrjoshi21 I guess that it does run on a background thread. And you cannot move Realm objects freely between threads.

**Comment by cmelchior** [comment]
This have been fixed by #2780 which was just merged, and will be part of next release (Either 1.0.2 or 1.1.0).
